### PR TITLE
Entity id bug

### PR DESCRIPTION
**Pull request by shans**
Note that this incorporates your changes too.

### CODE REVIEW

File: src/runtime/dom-particle-base.ts [hunk 1]
```diff
@@ -220,7 +220,7 @@ export class DomParticleBase extends Particle {
    */
   async appendRawDataToHandle(handleName: string, rawDataArray): Promise<void> {
     const handle = this.handles.get(handleName);
-    if (handle && handle.entityClass) {
+    if (handle && handle['entityClass']) {
```
Comment (csilvestrini) [inline src/runtime/dom-particle-base.ts]: I think checking for `handle && handle instanceof PreMutationEntityHandle` would be a bit clearer

File: src/runtime/particle.ts [hunk 2]
```diff
@@ -281,7 +281,8 @@ export class Particle {
   async setDescriptionPattern(connectionName: string, pattern): Promise<boolean> {
     const descriptions = this.handles.get('descriptions');
     if (descriptions) {
-      const entityClass = descriptions.entityClass;
+      // TODO(shanestephens): Fix this!
+      const entityClass = descriptions['entityClass'];
```
Comment (csilvestrini) [inline src/runtime/particle.ts]: Here too

File: src/runtime/storageNG/handle.ts [hunk 3]
```diff
@@ -118,22 +116,60 @@ export abstract class Handle<T extends CRDTTypeRecord> {
   }
 }
 
+abstract class PreEntityMutationHandle<T extends CRDTTypeRecord> extends Handle<T> {
```
Comment (csilvestrini) [inline src/runtime/storageNG/handle.ts]: Could you add a comment explaining what this class is for? Particularly addressing the meaning of "pre-mutation".

File: src/runtime/storageNG/handle.ts [hunk 4]
```diff
@@ -118,22 +116,60 @@ export abstract class Handle<T extends CRDTTypeRecord> {
...
+  entityClass: EntityClass;
+
+  constructor(key: string, storageProxy: StorageProxy<T>, idGenerator: IdGenerator,
+    particle: Particle, canRead: boolean, canWrite: boolean, name?: string) {
+      super(key, storageProxy, idGenerator, particle, canRead, canWrite, name);
```
Comment (csilvestrini) [inline src/runtime/storageNG/handle.ts]: Indenting here is weird. super should be indented once level, particle two.

File: src/runtime/storageNG/tests/handle-test.ts [hunk 5]
```diff
@@ -64,75 +70,109 @@ async function getSingletonHandle(particle?: MockParticle):
...
+let Bar: EntityClass;
+
+function newEntity(id: string) {
+  const bar = new Bar({});
+  Entity.identify(bar, id);
+  return bar;
+}
+
+async function containedIds(handle: CollectionHandle<Entity>): Promise<string[]> {
+  return (await handle.toList()).map(a => Entity.id(a));
```
Comment (csilvestrini) [inline src/runtime/storageNG/tests/handle-test.ts]: Just a simple FYI: you can return the promise here as-is, no need to await it.

Comment (shans) [reply src/runtime/storageNG/tests/handle-test.ts]: No, I need to map the conversion across the await'd list.

Comment (csilvestrini) [reply src/runtime/storageNG/tests/handle-test.ts]: Oh whoops, read that wrong

**Pull request merged**